File: src/truthound_dashboard/schemas/mask.py
"""Pydantic schemas for data masking (th.mask) operations.

Provides schemas for masking requests, responses, and history.
Supports three masking strategies: redact, hash, fake.
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Literal

from pydantic import Field

from .base import BaseSchema, IDMixin, TimestampMixin


class MaskingStrategy(str, Enum):
    """Masking strategy options.

    - redact: Replace values with asterisks (e.g., "john@example.com" -> "****")
    - hash: Replace values with SHA256 hash (deterministic, can be used for joins)
    - fake: Replace values with realistic fake data (e.g., "john@example.com" -> "alice@test.org")
    """

    REDACT = "redact"
    HASH = "hash"
    FAKE = "fake"


MaskingStrategyLiteral = Literal["redact", "hash", "fake"]


class MaskStatus(str, Enum):
    """Status of a masking operation."""

    PENDING = "pending"
    RUNNING = "running"
    SUCCESS = "success"
    FAILED = "failed"
    ERROR = "error"


class MaskRequest(BaseSchema):
    """Request body for running a masking operation.

    Attributes:
        columns: Optional list of columns to mask. If None, auto-detects PII.
        strategy: Masking strategy to use. Defaults to "redact".
        output_format: Output file format. Defaults to "csv".
    """

    columns: list[str] | None = Field(
        default=None,
        description="Columns to mask. If not specified, auto-detects PII columns.",
    )
    strategy: MaskingStrategyLiteral = Field(
        default="redact",
        description="Masking strategy: 'redact' (asterisks), 'hash' (SHA256), 'fake' (realistic data)",
    )
    output_format: Literal["csv", "parquet", "json"] = Field(
        default="csv",
        description="Output file format",
    )


class MaskSummary(BaseSchema):
    """Summary of a masking operation.

    Attributes:
        source_id: ID of the source that was masked.
        source_name: Name of the source.
        status: Current status of the operation.
        strategy: Masking strategy used.
        columns_masked: Number of columns that were masked.
        row_count: Number of rows processed.
        duration_ms: Operation duration in milliseconds.
    """

    source_id: str
    source_name: str | None = None
    status: str
    strategy: str
    columns_masked: int
    row_count: int | None = None
    duration_ms: int | None = None


class MaskResponse(BaseSchema, IDMixin, TimestampMixin):
    """Response for a masking operation.

    Attributes:
        id: Unique identifier for the masking operation.
        source_id: ID of the source that was masked.
        status: Current status (pending, running, success, failed, error).
        strategy: Masking strategy used.
        output_path: Path to the masked output file.
        columns_masked: List of columns that were masked.
        auto_detected: Whether PII columns were auto-detected.
        row_count: Number of rows processed.
        column_count: Total number of columns.
        duration_ms: Operation duration in milliseconds.
        error_message: Error message if operation failed.
        started_at: When the operation started.
        completed_at: When the operation completed.
    """

    source_id: str
    status: str
    strategy: str
    output_path: str | None = None
    columns_masked: list[str] | None = None
    auto_detected: bool = False
    row_count: int | None = None
    column_count: int | None = None
    duration_ms: int | None = None
    error_message: str | None = None
    started_at: datetime | None = None
    completed_at: datetime | None = None

    @classmethod
    def from_db(cls, db_mask: object) -> MaskResponse:
        """Create response from database model.

        Args:
            db_mask: DataMask database model instance.

        Returns:
            MaskResponse instance.
        """
        return cls(
            id=db_mask.id,
            source_id=db_mask.source_id,
            status=db_mask.status,
            strategy=db_mask.strategy,
            output_path=db_mask.output_path,
            columns_masked=db_mask.columns_masked,
            auto_detected=db_mask.auto_detected,
            row_count=db_mask.row_count,
            column_count=db_mask.column_count,
            duration_ms=db_mask.duration_ms,
            error_message=db_mask.error_message,
            started_at=db_mask.started_at,
            completed_at=db_mask.completed_at,
            created_at=db_mask.created_at,
            updated_at=getattr(db_mask, "updated_at", None),
        )


class MaskListItem(BaseSchema, IDMixin):
    """List item for masking operations.

    Attributes:
        id: Unique identifier.
        source_id: ID of the source.
        source_name: Name of the source.
        status: Current status.
        strategy: Masking strategy used.
        columns_masked: Number of columns masked.
        row_count: Number of rows processed.
        duration_ms: Operation duration in milliseconds.
        created_at: When the operation was created.
    """

    source_id: str
    source_name: str | None = None
    status: str
    strategy: str
    columns_masked: int = 0
    row_count: int | None = None
    duration_ms: int | None = None
    created_at: datetime

    @classmethod
    def from_db(cls, db_mask: object, source_name: str | None = None) -> MaskListItem:
        """Create list item from database model.

        Args:
            db_mask: DataMask database model instance.
            source_name: Optional source name.

        Returns:
            MaskListItem instance.
        """
        return cls(
            id=db_mask.id,
            source_id=db_mask.source_id,
            source_name=source_name or getattr(db_mask.source, "name", None),
            status=db_mask.status,
            strategy=db_mask.strategy,
            columns_masked=len(db_mask.columns_masked) if db_mask.columns_masked else 0,
            row_count=db_mask.row_count,
            duration_ms=db_mask.duration_ms,
            created_at=db_mask.created_at,
        )


class MaskListResponse(BaseSchema):
    """Response for listing masking operations.

    Attributes:
        data: List of masking operation items.
        total: Total number of items.
        limit: Maximum items per page.
    """

    data: list[MaskListItem]
    total: int
    limit: int = 20

File: frontend/src/mocks/handlers/mask.ts
/**
 * Data masking API handlers
 *
 * Supports th.mask() strategies:
 * - redact: Replace values with asterisks
 * - hash: Replace values with SHA256 hash
 * - fake: Replace values with realistic fake data
 */

import { http, HttpResponse, delay } from 'msw'
import { getStore, getById, create, getAll } from '../data/store'
import { createDataMask, createId } from '../factories'

const API_BASE = '/api/v1'

/**
 * Mask request options (mirrors MaskRequest from API)
 */
interface MaskRequest {
  columns?: string[]
  strategy?: 'redact' | 'hash' | 'fake'
  output_format?: 'csv' | 'parquet' | 'json'
}

export const maskHandlers = [
  // Run data masking for a source
  http.post(`${API_BASE}/masks/sources/:sourceId/mask`, async ({ params, request }) => {
    await delay(1500) // Simulate masking time

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    // Parse request body for mask options
    let options: MaskRequest = {}
    try {
      const body = await request.json()
      options = body as MaskRequest
    } catch {
      // Empty body is valid - use defaults
    }

    // Create new data mask operation
    const mask = createDataMask({
      id: createId(),
      sourceId,
      strategy: options.strategy ?? 'redact',
      columnsToMask: options.columns,
      autoDetected: !options.columns,
      status: 'success',
    })

    create(getStore().dataMasks, mask)

    return HttpResponse.json(mask, { status: 201 })
  }),

  // Get data mask by ID
  http.get(`${API_BASE}/masks/:id`, async ({ params }) => {
    await delay(150)

    const mask = getById(getStore().dataMasks, params.id as string)

    if (!mask) {
      return HttpResponse.json(
        { detail: 'Mask operation not found' },
        { status: 404 }
      )
    }

    return HttpResponse.json(mask)
  }),

  // List data masks for a source
  http.get(`${API_BASE}/masks/sources/:sourceId/masks`, async ({ params, request }) => {
    await delay(200)

    const url = new URL(request.url)
    const limit = parseInt(url.searchParams.get('limit') ?? '20')

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    const allMasks = getDataMasksBySourceId(sourceId)
    const masks = allMasks.slice(0, limit)

    return HttpResponse.json({
      data: masks.map((m) => ({
        id: m.id,
        source_id: m.source_id,
        source_name: source.name,
        status: m.status,
        strategy: m.strategy,
        columns_masked: m.columns_masked?.length ?? 0,
        row_count: m.row_count,
        duration_ms: m.duration_ms,
        created_at: m.created_at,
      })),
      total: allMasks.length,
      limit,
    })
  }),

  // Get latest data mask for a source
  http.get(`${API_BASE}/masks/sources/:sourceId/masks/latest`, async ({ params }) => {
    await delay(150)

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    const masks = getDataMasksBySourceId(sourceId)

    if (masks.length === 0) {
      return HttpResponse.json(
        { detail: 'No mask operations found for source' },
        { status: 404 }
      )
    }

    return HttpResponse.json(masks[0])
  }),
]

// Helper function to get data masks by source ID
function getDataMasksBySourceId(sourceId: string) {
  return getAll(getStore().dataMasks)
    .filter((m) => m.source_id === sourceId)
    .sort((a, b) => new Date(b.created_at).getTime() - new Date(a.created_at).getTime())
}
